project-koku/koku
# [COST-2231] Upgrade to Django 4

## Jira Ticket

[COST-2231](https://issues.redhat.com/browse/COST-2231)

## Description

This change will:

- [x] Upgrade Django from 3.2 to 4.0 and DRF to 3.14 - #4839 
- [x] Upgrade Django from 4.0 to 4.1 - #4853 
- [x] Upgrade Django from 4.1 to 4.2 - https://github.com/project-koku/koku/pull/4893



## Testing

1. Stop server containers
    ```
    docker compose stop koku-server masu-server worker
    ```

1. Update template schema and `clone_schema` SQL function
    ```
    python koku/manage.py migrate_serial_to_identity_columns --schema template0 --write
    ```
    
    Or update all schemas
    ```
    python koku/manage.py migrate_serial_to_identity_columns --write
    ```
    
1. Rebuild and start server contaiters
    ```
    make docker-up-min
    ```
    
1. Create test customer and load data. You may need to delete existing sources first.
    ```
    make create-test-customer
    make load-test-customer-data
    ```
    
1. Observe data processing
    ```
    make docker-logs
    ```
    
1. Verify sources were created and cost data was loaded
    ```
    http://localhost:8000/api/cost-management/v1/reports/oci/costs/
    http://localhost:8000/api/cost-management/v1/reports/azure/costs/
    http://localhost:8000/api/cost-management/v1/reports/aws/costs/
    # etc.
    ```
    
1. Verify template schema was correctly updated in database. Look for the ID column to be generated as identity. Do this on as many tables as you wish.
    ```
    psql --host $POSTGRES_SQL_SERVICE_HOST \
         --port $POSTGRES_SQL_SERVICE_PORT \
         --username $DATABASE_ADMIN \
         --password $DATABASE_PASSWORD
         
    \d template0.reporting_ocp_cost_category
    Table "template0.reporting_ocp_cost_category"
         Column     |  Type   | Collation | Nullable |             Default
    ----------------+---------+-----------+----------+----------------------------------
     id             | integer |           | not null | generated by default as identity
     ...
     ```
     
1. Verify the `clone_schema` SQL function was updated
    ```
    psql --host $POSTGRES_SQL_SERVICE_HOST \
     --port $POSTGRES_SQL_SERVICE_PORT \
     --username $DATABASE_ADMIN \
     --password $DATABASE_PASSWORD
     
     \sf clone_schema
     
     [compare the output to db_functions/clone_schema.sql]
     ```

1. Run tests
    ```
     tox -r -e py39
    ```
